### PR TITLE
#9486: Merge CCL line_all_gather to TTNN

### PR DESCRIPTION
**Pull request by Aswinmcw**
### Ticket
#9486 

### Problem description
CCL ops needs to be merged to TTNN

### What's changed
Merge CCL all_gather and line_all_gather to TTNN

### Checklist
- [x] All Post Commit : https://github.com/tenstorrent/tt-metal/actions/runs/9955435931
- [x] T3000 Frequent Test : https://github.com/tenstorrent/tt-metal/actions/runs/9954388528
- [x] T3000 Model Perf Test : https://github.com/tenstorrent/tt-metal/actions/runs/9954390484
- [x] T3000 Profiler Test : https://github.com/tenstorrent/tt-metal/actions/runs/9955373842
- [x] T3000 Unit Test : https://github.com/tenstorrent/tt-metal/actions/runs/9954396008
- [x] T3000 Demo Test : https://github.com/tenstorrent/tt-metal/actions/runs/9954392283

### CODE REVIEW

**Review by SeanNijjar — COMMENTED** [review]
Thanks Aswin, this change looks good to me based on the diff.

Please make sure to run post-commit and t3000 frequent workflows to ensure there are no regressions:
https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml
https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-frequent-tests.yaml

File: ttnn/cpp/ttnn/operations/ccl.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl.hpp:41]: this should be moved to a separate file

The expected structure is 

operations/ccl/line_all_gather <-- files related to this op go here
operations/ccl/all_gather
operations/ccl/reduce_scatter

File: tt_eager/tt_dnn/op_library/all_gather/all_gather_op.cpp
Comment (ayerofieiev-tt) [inline tt_eager/tt_dnn/op_library/all_gather/all_gather_op.cpp:187]: when will this move to ttnn? so far this PR does not deliver to its title

**Review by TT-BrianLiu — APPROVED** [review]

**Review by eyonland — CHANGES_REQUESTED** [review]
Where is the code to register the op?

**Review by eyonland — CHANGES_REQUESTED** [review]
Can we move the unit test to ttnn unit tests?

**Comment by eyonland** [comment]
> Where is the code to register the op?

I see it... please ignore.

**Comment by SeanNijjar** [comment]
> Can we move the unit test to ttnn unit tests?

Hey Eyon, which unit tests are you referring to here? Are you referring to the pytest (test_all_gather.py) or the ccl gtests?
If the former, I agree - we should move them to the appropriate place in the new folder structure (Is that in `tests/ttnn/unit_tests/operations/`?)

If the latter, those tests are op independent and (generally) need to be moved as well, along with gtests. I'm thinking we can make that a separate task to do that move when, or after, the last ccl op is migrated (looks like it'll be reduce scatter).

**Comment by Aswinmcw** [comment]
Hi @eyonland @SeanNijjar , I have moved the pytest from tt eager to ttnn (`tests/ttnn/unit_tests/operations`), and during merging of next op (ie. reduce scatter), i will move the gtest to the TTNN

File: tests/scripts/run_tt_eager.py
Comment (SeanNijjar) [inline tests/scripts/run_tt_eager.py:35]: Hey @Aswinmcw, where did these go?

**Comment by mo-tenstorrent** [comment]
Can you make sure this https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-profiler-tests.yaml passes

**Comment by SeanNijjar** [comment]
> Can you make sure this https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-profiler-tests.yaml passes

It will also be easier if the pipeline runs are linked to here as well (either as a comment or in the PR description

File: tests/scripts/run_tt_eager.py
Comment (Aswinmcw) [reply tests/scripts/run_tt_eager.py]: Hi @SeanNijjar , i have reverted these gtest changes and only moved the pytest to TTNN unit test and modified the file paths in those. Kindly review it once

**Comment by SeanNijjar** [comment]
@Aswinmcw if you can resolve the conflict, then I  can approve this pending successful runs of the listed pipelines. I think @eyonland's comment is addressed now too.

**Review by tt-aho — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Review by SeanNijjar — APPROVED** [review]

**Review by mo-tenstorrent — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_all_gather.py
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_all_gather.py:233]: can you leave comments on whats wrong with those tests, why you had to disable?

@SeanNijjar on your radar?

File: tests/ttnn/unit_tests/operations/test_all_gather.py
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_all_gather.py:566]: why?

File: tests/ttnn/unit_tests/operations/test_all_gather.py
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_all_gather.py:887]: why?

File: ttnn/cpp/pybind11/operations/__init__.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/pybind11/operations/__init__.hpp:59]: Why? :) 

Please use single submodule ccl, no need for separate submodules for all_gather and line_all_gather.

File: ttnn/cpp/ttnn/operations/ccl/all_gather/ccl_all_gather_pybind.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/all_gather/ccl_all_gather_pybind.hpp:18]: just `ccl`

File: ttnn/cpp/ttnn/operations/ccl/line_all_gather/ccl_line_all_gather_pybind.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/line_all_gather/ccl_line_all_gather_pybind.hpp:18]: just ccl is good

File: ttnn/cpp/ttnn/operations/ccl/line_all_gather/ccl_line_all_gather_pybind.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/line_all_gather/ccl_line_all_gather_pybind.hpp:23]: bind_line_all_gather

**Review by SeanNijjar — CHANGES_REQUESTED** [review]
Undo test case disables.

File: ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/line_all_gather_op.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/line_all_gather_op.cpp:126]: why do we need line_all_gather and line_all_gather_impl and why do we want them here? this usually would belong to `line_all_gather.cpp` insde execute_on_work_thread

File: ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/line_all_gather_op.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/line_all_gather_op.hpp:31]: please avoid usings in headers as much as possible. here what you did is exposed this symbol inside ttnn::utils namespace, which seems fairly large

File: ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/line_all_gather_op.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/line_all_gather_op.hpp:62]: afaik, these can be removed after 
https://github.com/tenstorrent/tt-metal/pull/9994

**Comment by Aswinmcw** [comment]
> Undo test case disables.

@SeanNijjar I am getting the same error for the disabled cases as they mentioned in the issue #10086 , can i enable those cases for now?

**Comment by SeanNijjar** [comment]
> > Undo test case disables.
> 
> @SeanNijjar I am getting the same error for the disabled cases as they mentioned in the issue #10086 , can i enable those cases for now?

Yes. None of the tests should be disabled. I was only seeing one of the all-gather post commit tests fail on main so I would only expect that one to fail in your frequent run.

File: tests/ttnn/unit_tests/operations/test_all_gather.py
Comment (cfjchu) [inline tests/ttnn/unit_tests/operations/test_all_gather.py:413]: Can we start moving away from the old-style way of dealing with these multi-device tensors?

For example:
```
    input_tensors = torch.chunk(input_tensor, num_devices, dim)
    tt_input_tensors = []
    for i, t in enumerate(input_tensors):
        tt_input_tensors.append(ttl.tensor.Tensor(t, input_dtype).to(layout).to(devices[i], mem_config))

    input_tensor_mesh = ttnn.aggregate_as_tensor(tt_input_tensors)
```
can be replaced with:
```
ttnn_tensor = ttnn.from_torch(full_tensor, mesh_mapper=ShardTensorToMesh(device_mesh, dim=dim))
```


Take a look at: `tests/ttnn/unit_tests/test_multi_device.py::test_ttnn_multi_device_all_gather`

File: ttnn/cpp/ttnn/operations/ccl/shared_with_host/hetergeneous_data_structs.hpp
Comment (cfjchu) [inline ttnn/cpp/ttnn/operations/ccl/shared_with_host/hetergeneous_data_structs.hpp:307]: fix this

Comment (Aswinmcw) [reply ttnn/cpp/ttnn/operations/ccl/shared_with_host/hetergeneous_data_structs.hpp]: Fixed

File: tests/ttnn/unit_tests/operations/test_all_gather.py
Comment (cfjchu) [inline tests/ttnn/unit_tests/operations/test_all_gather.py:365]: We're losing coverage here.. `pcie_device_mesh` is a fixture that only returns the PCIE devices. You probably want to swap this with `t3k_device_mesh`

File: tests/ttnn/unit_tests/operations/test_all_gather.py
Comment (cfjchu) [inline tests/ttnn/unit_tests/operations/test_all_gather.py:381]: you should be able to do:

```
for device in pcie_device_mesh.get_devices():
    device.enable_async(enable_async)
```

Comment (Aswinmcw) [reply tests/ttnn/unit_tests/operations/test_all_gather.py]: Done

File: tests/ttnn/unit_tests/operations/test_all_gather.py
Comment (Aswinmcw) [reply tests/ttnn/unit_tests/operations/test_all_gather.py]: Updated the code to use t3k_device_mesh

File: ttnn/cpp/ttnn/operations/ccl/ccl_common.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/ccl_common.cpp:95]: kernels into a kernels folder please `ttnn/cpp/ttnn/operations/ccl/kernels/edm/erisc_datamover.cpp`

File: ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/ccl_line_all_gather_op.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/ccl_line_all_gather_op.hpp:14]: this file should be in line_all_gather folder, one level up and should be called `line_all_gather.hpp`

File: ttnn/cpp/ttnn/operations/ccl/line_all_gather/ccl_line_all_gather_pybind.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/ccl/line_all_gather/ccl_line_all_gather_pybind.hpp:23]: this file should be called `line_all_gather_pybind.hpp`

File: ttnn/cpp/pybind11/operations/__init__.hpp
Comment (cfjchu) [inline ttnn/cpp/pybind11/operations/__init__.hpp:60]: why are these separate python modules instead of single `ccl`?

Comment (Aswinmcw) [reply ttnn/cpp/pybind11/operations/__init__.hpp:60]: Both are declared in two different files and if I use same name it is showing conflict so I just defined two modules for it

**Comment by SeanNijjar** [comment]
@Aswinmcw, The fix for the t3000 allgather regression has been mainlined. Can you please rebase and kick off new a new t3000 frequent pipeline after the questions from @cfjchu and @ayerofieiev-tt have been addressed?

File: ttnn/cpp/ttnn/operations/ccl/ccl_common.cpp
Comment (Aswinmcw) [reply ttnn/cpp/ttnn/operations/ccl/ccl_common.cpp]: @ayerofieiev-tt I have moved the files to kernels directory

**Review by SeanNijjar — APPROVED** [review]

**Comment by SeanNijjar** [comment]
Hey @Aswinmcw, can you please also run this pipeline https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml

The pipeline itself is not green at the moment but we can make sure we aren't regressing the passing tests or causing the failing test to fail differently.

File: tests/scripts/run_profiler_regressions.sh
Comment (ayerofieiev-tt) [inline tests/scripts/run_profiler_regressions.sh:8]: Going forward please do `if grayskull return` in such cases so the rest of the code does not have to be that indentended.

File: tests/tt_eager/ops/ccl/test_all_gather_sharded_indexing_helpers.cpp
Comment (ayerofieiev-tt) [inline tests/tt_eager/ops/ccl/test_all_gather_sharded_indexing_helpers.cpp:149]: I intuitively expect this to be `ttnn::ccl::all_gather::utils::..`

File: tt_eager/tt_dnn/op_library/ccl/reduce_scatter/kernels/worker_interleaved_ring_reduce_scatter_reader.cpp
Comment (ayerofieiev-tt) [inline tt_eager/tt_dnn/op_library/ccl/reduce_scatter/kernels/worker_interleaved_ring_reduce_scatter_reader.cpp:37]: and this `ttnn::ccl::utils::..`

File: tt_eager/tt_dnn/op_library/ccl/reduce_scatter/reduce_scatter_op.cpp
Comment (ayerofieiev-tt) [inline tt_eager/tt_dnn/op_library/ccl/reduce_scatter/reduce_scatter_op.cpp:75]: not clear why this is in utils namespace.. looks like it would be better to just remove the utils namespace and keep things in `ttnn::ccl::..`

File: ttnn/cpp/pybind11/operations/__init__.hpp
Comment (ayerofieiev-tt) [reply ttnn/cpp/pybind11/operations/__init__.hpp:60]: these are not modules anymore, would be good to rename to smth like `ccl::py_bind_all_gather`

File: ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/line_all_gather_op.cpp
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/ccl/line_all_gather/device/line_all_gather_op.cpp:126]: @VirdhatchaniKN

**Review by ayerofieiev-tt — APPROVED** [review]
I see some issues with 
* utils namespace
* names of binding functions
* and location of op call logic (micro op and call logic look mixed in one place)

None of these are hard to change and none of them are blockers.
I would recommend to fire a separate ticket.

File: tests/scripts/run_profiler_regressions.sh
Comment (ttmchiou) [reply tests/scripts/run_profiler_regressions.sh:8]: Would this not be better if we have this as "ARCH_NAME" = wormhole_b0?
This script would trigger if we introduce another ARCH in the future

Comment (Aswinmcw) [reply tests/scripts/run_profiler_regressions.sh:8]: Okay I will modify it to trigger only for wormhole_b0 instead of skipping it for GS

Comment (ttmchiou) [reply tests/scripts/run_profiler_regressions.sh:8]: can you also update the same reference in this file too?
I believe there is another reference doing the same thing  for run_async_mode_T3000_test

Comment (mo-tenstorrent) [reply tests/scripts/run_profiler_regressions.sh:8]: Great idea, please make sure. 
https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-profiler-tests.yaml after the change. 

https://github.com/tenstorrent/tt-metal/actions/workflows/run-profiler-regression.yaml should also pass as part of the post commit change.

File: tests/scripts/run_profiler_regressions.sh
Comment (ttmchiou) [inline tests/scripts/run_profiler_regressions.sh:8]: I believe this would still create an issue like the first version of code.
If future new ARCH_NAMEs are release which was not compatible with this test, we would still encounter the same scenario where this this test is unintentionally executed and 
Since t3000s are running on wormhole_b0 cards, it would be better to explicitly set it as execute for only wormhole_b0 cards.
```
if [ "$ARCH_NAME" = "wormhole_b0" ]; then
    logic
fi
```

**Review by ttmchiou — APPROVED** [review]

**Comment by ttmchiou** [comment]
make sure new changes are tested too before merging

File: tests/scripts/run_profiler_regressions.sh
Comment (tt-rkim) [inline tests/scripts/run_profiler_regressions.sh:8]: We don't allow such if statements anymore

Please skip the test using fixtures

Comment (mo-tenstorrent) [reply tests/scripts/run_profiler_regressions.sh:8]: All for not allowing these ifs, are the pytest skip fixtures in?

**Review by tt-rkim — APPROVED** [review]

**Pull request merged**
